packages/pouch-storage/src/PouchCommandHistoryTable.ts

```diff
@@ -24,6 +37,55 @@ export class PouchCommandHistoryTable
         : []
     );
   }
+
+  dbUpdate(
+    event: PouchDB.Core.ChangesResponseChange<CommandHistoryStorageItem>
+  ): void {
+    log.debug('Clearing cache and refreshing data', event);
+
+    this.cache = null;
+
+    super.dbUpdate(event);
+  }
+
+  /**
+   * Fetch command history storage items.
+   * This override provides caching + optimizations specific to command history.
+   *
+   * 1. The table doesn't support sorting, so ignore the sort param.
+   * 2. We cache the results to avoid excessive PouchDB queries during scrolling.
+   * The cache should only need to be invalidated when search filter changes or
+   * commands are added to the storage history.
+   *
+   * @param viewport
+   * @param selector
+   * @param _sort
+   * @returns Promise to array of command history storage items
+   */
+  async fetchViewportData(
+    viewport: StorageTableViewport,
+    selector: PouchDB.Find.Selector,
+    _sort: PouchDBSort
+  ): Promise<ViewportData<CommandHistoryStorageItem>> {
+    if (this.cache) {
+      log.debug('Fetching from cache', viewport, selector);
+    } else {
+      log.debug('Fetching from PouchDB', selector);
+
+      this.cache = await this.db
+        .find({
+          selector,
+          fields: ['id', 'name'],
```

Comment:
We're not setting a viewport anymore, just fetching it all at once? How large can this grow/what kind of stress testing did you do? At least deserves a comment here...
I'm hesitant this is unbounded. What happens when someone has a million items in their history? We should at least truncate it at a limit if we're going to do this.

Reply:
This might be more the tree to bark up for the unfiltered view.

https://pouchdb.com/guides/bulk-operations.html#please-use-alldocs

> Far too many developers overlook this valuable API, because they misunderstand it. When a developer says "my PouchDB app is slow!", it is usually because they are using the slow query() API when they should be using the fast allDocs() API.
> 
> For details on how to effectively use allDocs(), you are strongly recommended to read ["Pagination strategies with PouchDB"](http://pouchdb.com/2014/04/14/pagination-strategies-with-pouchdb.html). For 99% of your applications, you should be able to use allDocs() for all the pagination/sorting/searching functionality that you need.

Reply:
I committed a first pass at fixing the performance using allDocs for non-filtered queries. Also addressed some other issues where we were hitting the db multiple times a second. Tested with 20,000 items in storage. Very speedy when no search applied. A bit of a delay on initial search, but then scrolling is fine. Still need to do some more testing / tweaking.